src/Guide/Api/Utils.hs

```diff
@@ -89,3 +94,43 @@ instance (HasSwagger api, KnownSymbol name, KnownSymbol desc) =>
              symbolVal (Proxy @desc))
             Nothing
      in toSwagger (Proxy @api) & applyTags [tag]
+
+
+-- | Type to log user's edition with header details
+data RequestDetails = RequestDetails
+  { rdIp      :: Text
+  , rdReferer :: Text
+  , rdUA      :: Text
+  } deriving (Show, Generic)
+
+instance (HasServer api context)
+  => HasServer (Maybe RequestDetails :> api) context where
+
+  type ServerT (Maybe RequestDetails :> api) m = Maybe RequestDetails -> ServerT api m
+
+  hoistServerWithContext _ pc nt s = hoistServerWithContext (Proxy :: Proxy api) pc nt . s
+
+  route Proxy context subserver = route (Proxy :: Proxy api) context $
+      subserver `addHeaderCheck` withRequest headerCheck
+    where
+      headerCheck :: Request -> DelayedIO (Maybe RequestDetails)
+      headerCheck req = unfoldRequestArgument (Proxy :: Proxy '[Optional, Strict]) errReq errSt meRequestDatails
+        where
+          meRequestDatails :: Maybe (Either Text RequestDetails)
+          meRequestDatails = getCompose $ liftA2 RequestDetails (Compose ip) (Compose referer) <*> Compose ua
+
+          ip :: FromHttpApiData a => Maybe (Either Text a)
+          ip = fmap parseHeader $ lookup "X-Forwarded-For" (requestHeaders req)
+          referer :: FromHttpApiData a => Maybe (Either Text a)
+          referer = fmap parseHeader $ lookup hReferer (requestHeaders req)
+          ua :: FromHttpApiData a => Maybe (Either Text a)
+          ua = fmap parseHeader $ lookup hUserAgent (requestHeaders req)
+
+          errReq = delayedFailFatal err400
+            { errBody = "One of Headers: " <> "Referer, X-Forwarded-For or User-Agent" <> " is required" }
```

Comment:
Is it a case for 404? There are no like "Not found", it is just a validation error and it should be 400

Reply:
It shouldn't even be 400, it should just result in a `Nothing`. We definitely _definitely_ don't want to throw errors of any kind for requests that don't have a `Referer`. This will break curl, this will break people who intentionally hide the referer, etc.